Lib/dataclasses.py

```diff
@@ -1293,11 +1316,22 @@ class C:
 
 def _asdict_inner(obj, dict_factory):
     if _is_dataclass_instance(obj):
-        result = []
-        for f in fields(obj):
-            value = _asdict_inner(getattr(obj, f.name), dict_factory)
-            result.append((f.name, value))
-        return dict_factory(result)
+        if dict_factory is dict:
+            result = {}
+            for f in fields(obj):
+                value = getattr(obj, f.name)
+                if type(value) not in _ATOMIC_TYPES:
+                    value = _asdict_inner(value, dict_factory)
+                result[f.name] = value
+            return result
```

Comment:
Add a comment explaining why we have this branch at all (optimising for the common case):

```suggestion
            # Most of the time dict_factory will be dict,
            # so optimize for the common case: 
            result = {}
            for f in fields(obj):
                value = getattr(obj, f.name)
                if type(value) not in _ATOMIC_TYPES:
                    value = _asdict_inner(value, dict_factory)
                result[f.name] = value
            return result
```

Reply:
This discussion happened in a separate thread - but there will be a separate PR for this as agreed.